### PR TITLE
Add test machinery for OpenTelemetry

### LINKED ISSUE #4323: Add integration tests to cover OpenTelemetry stack integration
Add the tests and the machinery needed to cover the OpenTelemetry stack integration.

This should cover the traffic between the instrumented code and the collector that collects trace and metrics spans.

### PR DESCRIPTION
**Pull request by lubosmj**
Tests checking OpenTelemetry payloads are now run in serial.

closes #4323

credits to @mdellweg for designing the machinery

### CODE REVIEW

File: pulpcore/tests/functional/__init__.py [hunk 1]
```diff
@@ -413,10 +416,104 @@ def make_url(self, path):
...
+    collector_server.join()
+    if expected_spans:
+        raise Exception(
+            f"Expected spans are still missing: {expected_spans}\n"
+            f"Unexpeced spans received: {unexpected_spans}"
+        )
 
 
 @pytest.fixture
+def expect_span(_otel_collector):
```
Comment (mdellweg) [inline pulpcore/tests/functional/__init__.py:505]: Should we call it expect_otel_span ?

File: .github/workflows/scripts/post_before_install.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+if [ "$TEST" = "azure" ]; then
+    # configuration variables required by the opentelemetry tests
+    export OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
+    export OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
+    export OTEL_EXPORTER_OTLP_TIMEOUT=1
+    export OTEL_BSP_SCHEDULE_DELAY=10
+    export OTEL_BSP_EXPORT_TIMEOUT=10
+    export OTEL_BSP_MAX_QUEUE_SIZE=1
+    export OTEL_BSP_MAX_EXPORT_BATCH_SIZE=1
```
[the remark this answers is not in the document]
Comment (lubosmj) [reply .github/workflows/scripts/post_before_install.sh]: Yes, at least, it worked locally. We need to incorporate these variables into the CI.

**Comment by lubosmj** [comment]
How big of a deal is to run the status tests in serial? There might be a case where the same code for running the collector server is executed from multiple processes simultaneously, resulting in "address already in use" errors. Maybe, because of the following issue: https://stackoverflow.com/questions/61216175/how-to-run-setup-fixture-once-and-then-tests-in-parallel.

In the near future, we can end up running more and more tests in serial because of this issue. For now, is it possible to run a separate Python server next to the pytest process? The server will represent a standalone otel collector that will be tickled (by sending otel data from Pulp, triggered by tests) and queried (by retrieving otel data back, used for test assertions) by the pytest process. Thus, the server will act as a temporary storage for otel data.

**Comment by lubosmj** [comment]
Should we worry about these errors in the logs? They are not harmful, see the azure runner.
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/opentelemetry/sdk/metrics/_internal/export/__init__.py", line 523, in _receive_metrics
    self._exporter.export(
  File "/usr/local/lib/python3.8/site-packages/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py", line 189, in export
    resp = self._export(serialized_data.SerializeToString())
  File "/usr/local/lib/python3.8/site-packages/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py", line 162, in _export
    return self._session.post(
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 637, in post
    return self.request("POST", url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/adapters.py", line 519, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=4318): Max retries exceeded with url: /v1/metrics (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fd6ddfd6f70>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

These errors are raised when otel data cannot reach the destination collector point. The collector server is initialized only when running `test_status` tests. However, the otel data are sent to the collector throughout the whole testing process.

File: .github/workflows/scripts/post_before_install.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (lubosmj) [reply .github/workflows/scripts/post_before_install.sh]: I have managed to scale down the number of required settings. Thanks, @decko, for finding these.

File: .ci/ansible/Containerfile.j2 [hunk 3]
```diff
@@ -30,6 +30,14 @@ RUN pip3 install
 {%- endif -%}
 {%- endfor %}
 
+{% if azure_test | default(false) %}
+ENV PULP_OTEL_ENABLED=true
+ENV OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
+ENV OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
+ENV OTEL_BSP_MAX_QUEUE_SIZE=1
+ENV OTEL_BSP_MAX_EXPORT_BATCH_SIZE=1
+{% endif %}
```
Comment (lubosmj) [inline .ci/ansible/Containerfile.j2:39]: This gets overwritten when re-applying the plugin template. How should we tackle that? I think the variables need to be set before running `/init`.

Comment (mdellweg) [reply .ci/ansible/Containerfile.j2:39]: Can we use the plugin template to inject scenario specific env-vars? In a similar way we can currently inject settings. And yes, env-vars would superseed settings by the power of dynaconf. XD

Comment (lubosmj) [reply .ci/ansible/Containerfile.j2]: Done in https://github.com/pulp/plugin_template/pull/786.

**Comment by lubosmj** [comment]
The tests seem to be working because I could not find "Telemetry tests are disabled" warnings in the logs. 🐕‍🦺

**Comment by decko** [comment]
> Should we worry about these errors in the logs? They are not harmful, see the azure runner.
> 
> ```
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.8/site-packages/opentelemetry/sdk/metrics/_internal/export/__init__.py", line 523, in _receive_metrics
>     self._exporter.export(
>   File "/usr/local/lib/python3.8/site-packages/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py", line 189, in export
>     resp = self._export(serialized_data.SerializeToString())
>   File "/usr/local/lib/python3.8/site-packages/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py", line 162, in _export
>     return self._session.post(
>   File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 637, in post
>     return self.request("POST", url, data=data, json=json, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 589, in request
>     resp = self.send(prep, **send_kwargs)
>   File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 703, in send
>     r = adapter.send(request, **kwargs)
>   File "/usr/local/lib/python3.8/site-packages/requests/adapters.py", line 519, in send
>     raise ConnectionError(e, request=request)
> requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=4318): Max retries exceeded with url: /v1/metrics (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fd6ddfd6f70>: Failed to establish a new connection: [Errno 111] Connection refused'))
> ```
> 
> These errors are raised when otel data cannot reach the destination collector point. The collector server is initialized only when running `test_status` tests. However, the otel data are sent to the collector throughout the whole testing process.

My only concern is that we get lots of noise/garbage that make the test logs useless.

File: pulpcore/tests/functional/__init__.py [hunk 4]
```diff
@@ -413,10 +416,98 @@ def make_url(self, path):
...
+    # Not sure what the right time to wait is here but the server does not seem
+    # to be found immediately.
+    for _ in range(30):
+        if unexpected_spans:
+            # Reset unexpected spans
+            unexpected_spans = []
+            break
+        # Tickle the telemetry stack
+        status_api_client.status_read()
+        time.sleep(1)
```
Comment (decko) [inline pulpcore/tests/functional/__init__.py:483]: With this, won't we be generating 30 spans?
If it's about waiting for something, could we just use time.sleep(30)?

Comment (lubosmj) [reply pulpcore/tests/functional/__init__.py:483]: We might generate 30 spans when the server is not up and ready for 30 seconds. We exit the loop immediately if we receive any span from "tickling". That means the server is running and receiving data.

**Comment by lubosmj** [comment]
> My only concern is that we get lots of noise/garbage that make the test logs useless.

Good point. Is there any way to enable the telemetry just for the status endpoint? Would using `autouse=True` in the otel server fixture work for us?

**Comment by mdellweg** [comment]
> > My only concern is that we get lots of noise/garbage that make the test logs useless.
> 
> Good point. Is there any way to enable the telemetry just for the status endpoint? Would using `autouse=True` in the otel server fixture work for us?

We want to assert telemetry on many other tests too, right? The status endpoint was just the guinea pig.

**Comment by lubosmj** [comment]
@decko, @mdellweg, having defined `autouse=True` for the otel collector fixture removed the errors almost completely. There are still a couple of them at the beginning and end of the test run (for instance, when the fixture is not yet ready or was destroyed but pulp still sends some bits of data). What do you think?

**Comment by lubosmj** [comment]
Three proposals:
1. EASY. Leave the code as it is, remove `autouse=True`, and ignore the fact that we flood logs with "Connection refused".
2. MEDIUM. Monkeypatch `opentelemetry-python` from tests to skip logging "Connection Refused" errors (https://github.com/open-telemetry/opentelemetry-python/blob/0c382ee5356e1fa5a292e7a6c4961390cdfbe6df/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L368). *I am now working with this proposal.*
3. MEDIUM. Run a standalone HTTP server process that will run in the background and act as a storage for otel data.

**Comment by lubosmj** [comment]
I was not able to monkeypatch the otel SDK.

But, running the otel collector server in a separate process before executing all tests worked for me. The tracebacks are almost gone. Only a couple of them are shown in the logs as Pulp starts up and the server is not yet up and running (because it is dispatched from the `before_script` script).

File: pulpcore/tests/functional/otel_server.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
+import asyncio
```
Comment (mdellweg) [inline pulpcore/tests/functional/otel_server.py:1]: Nice work!
I would have preferred to run this as a thread in the pytest process, but we want something working now.
Could this live in `pulpcore/tests/functional/assets` or similar?

**Review by mdellweg — APPROVED** [review]
Let's use this approach and see how far we get.

File: pulpcore/tests/functional/otel_server.py (hunk 5 above)
Comment (lubosmj) [reply pulpcore/tests/functional/otel_server.py:1]: Running this in the thread is very tricky. We want to have the server running always to not spam the logs with "connection refused". At the same time, we want to access the data from the server to enable assertions from tests. If the tests are run in parallel, we do not know which worker will pick a specific test even if we ensure that the server fixture is run from gw0. Thus, there is no guarantee we can access the server's data eventually.

Comment (mdellweg) [reply pulpcore/tests/functional/otel_server.py]: Yes, that is not manageable on the fixture level, but maybe as a pytest extension somehow. But also not something we need to experiment with today.

**Review by decko — APPROVED** [review]
I think it's a GO.
Great work @lubosmj

**Pull request merged**